namespace_validate.go

```diff
@@ -0,0 +1,118 @@
+package kubeauth
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	k8s_yaml "k8s.io/apimachinery/pkg/util/yaml"
+)
+
+// This exists so we can use a mock namespace validation when running tests
+type namespaceValidator interface {
+	ValidateLabels(context.Context, *http.Client, string, string) (bool, error)
+}
+
+type namespaceValidateFactory func(*kubeConfig) namespaceValidator
+
+// This is the real implementation that calls the kubernetes API
+type namespaceValidateAPI struct {
+	config *kubeConfig
+}
+
+func namespaceValidateAPIFactory(config *kubeConfig) namespaceValidator {
+	return &namespaceValidateAPI{
+		config: config,
+	}
+}
+
+func (v *namespaceValidateAPI) ValidateLabels(ctx context.Context, client *http.Client, namespace string, namespaceSelector string) (bool, error) {
+	labelSelector, err := makeLabelSelector(namespaceSelector)
+	if err != nil {
+		return false, err
+	}
+	nsLabels, err := v.getNamespaceLabels(ctx, client, namespace)
+	if err != nil {
+		return false, err
+	}
+	selector, err := metav1.LabelSelectorAsSelector(&labelSelector)
+	if err != nil {
+		return false, err
+	}
+	return selector.Matches(labels.Set(nsLabels)), nil
+}
+
+func makeLabelSelector(selector string) (metav1.LabelSelector, error) {
+	labelSelector := metav1.LabelSelector{}
+	decoder := k8s_yaml.NewYAMLOrJSONDecoder(strings.NewReader(selector), len(selector))
+	err := decoder.Decode(&labelSelector)
+	if err != nil {
+		return labelSelector, err
+	}
+	return labelSelector, nil
+}
+
+func (v *namespaceValidateAPI) getNamespaceLabels(ctx context.Context, client *http.Client, namespace string) (map[string]string, error) {
```

Comment:
I wonder if it make things simpler if we starting the k8s client instead of this homebrew client. We should consider that for a future release.

Reply:
I'm leaning towards not using k8s client based on the [above discussion](https://github.com/hashicorp/vault-plugin-auth-kubernetes/pull/218#discussion_r1407723039) for now.